src/coreclr/debug/daccess/enummem.cpp

```diff
@@ -1582,6 +1585,57 @@ HRESULT ClrDataAccess::EnumMemCLRMainModuleInfo()
     return status;
 }
 
+HRESULT ClrDataAccess::EnumMemCollectFrozenHeap()
+{
+    SUPPORTS_DAC;
+
+    HRESULT hr = S_OK;
+
+    struct DacpGcHeapData heap;
+    IfFailRet(GetGCHeapData(&heap));
+    if (heap.bServerMode)
+    {
+        unsigned int heapCount = GCHeapCount();
+        CLRDATA_ADDRESS* heaps = (CLRDATA_ADDRESS*)alloca(heapCount * sizeof(CLRDATA_ADDRESS*));
+        IfFailRet(GetServerHeaps(heaps, m_pTarget));
+        for (unsigned int heapNum = 0; heapNum < heapCount; heapNum++)
+        {
+            struct DacpGcHeapDetails heap_detail;
+            IfFailRet(ServerGCHeapDetails(heaps[heapNum], &heap_detail));
+            IfFailRet(EnumMemCollectFrozenSegments(heap_detail.generation_table[2].start_segment));
+        }
+    }
+    else
+    {
+        struct DacpGcHeapDetails heap_detail;
+        IfFailRet(GetGCHeapStaticData(&heap_detail));
+        IfFailRet(EnumMemCollectFrozenSegments(heap_detail.generation_table[2].start_segment));
+    }
+
+    return hr;
+}
+
+HRESULT ClrDataAccess::EnumMemCollectFrozenSegments(CLRDATA_ADDRESS segment)
+{
+    SUPPORTS_DAC;
+
+    HRESULT hr = S_OK;
+
+    const size_t heap_segment_flags_readonly = 1;
+
+    while (segment)
+    {
+        struct DacpHeapSegmentData heap_segment;
+        IfFailRet(GetHeapSegmentData(segment, &heap_segment));
+        if (heap_segment.flags & heap_segment_flags_readonly)
```

Comment:
Any reason EnumWksGlobalMemoryRegions doesn't already cover this case? And if there's a reason it doesn't work, can it be added to those functions?

Reply:
It looks like the innermost call is only enumerating the memory for the `heap_segment` object describing the segment but not the segment itself.

`DacEnumMemoryRegion(dac_cast<TADDR>(seg), sizeof(dac_heap_segment));`

Note that `sizeof(dac_heap_segment)`, that's a tiny number compared to the whole segment.

Reply:
Yeah, it's the tracking. Feels off to have to go through these in this manner to find the frozen segments. In some way I like that they are treated like segments, but since you pointed out they are now executableAllocator reserved, the following happens:

- ~Why don't they get collected with the rest of the ExecutableAllocator pages?~ Looks like other uses are on `UnlockedLoaderHeap`, `HostCodeHeap` cases that use the ExecutableAllocator, but they themselves enumerate the memory. The only other ones that I don't see reported is `LoaderAllocator::m_InitialReservedMemForLoaderHeaps` and `VirtualCallStubManager::m_initialReservedMemForHeaps` which changed with W^X and might also need some investigation. It seems like they get reported in an ad-hoc more granular manner. This could be an issue by itself... For example calls to `ReserveMemoryFromExecutableAllocator` might not be reported - not an issue right now since these are for image ranges that we seem to get from maps.
- Would it be a good idea to teach the `FrozenObjectHeapManager` to enumerate memory and that way report `FrozenObjectSegment`s? What's allocated in this memory, since it might be relevant for triage dump scenarios?

Reply:
@EgorBo for `FrozenObjectHeapManager`
@janvorli for `ExecutableAllocator`

Reply:
> ReserveMemoryFromExecutableAllocator

This is used only by the PAL code that maps PE files into memory. The code uses this or `mmap` (in case there is not enough memory left in the PAL preallocated virtual memory range or the OS is 32 bit). Runtime calls `PAL_LOADLoadPEFile` that invokes that code. On Windows, it calls `CLRLoadLibraryEx` instead. 
@hoyosjs should we add logging of loaded PE binaries to DAC code?

Reply:
> `LoaderAllocator::m_InitialReservedMemForLoaderHeaps` and `VirtualCallStubManager::m_initialReservedMemForHeaps`

It seems these should not need reporting. They only reserve virtual memory range, the memory is committed by the users of those ranges (`LoaderHeap`).

Reply:
The revert is https://github.com/dotnet/runtime/pull/80400 in case we feel that's the more immediate solution.

Reply:
Since the revert is merged, I will close this one.